File: src/main/java/fi/aalto/cs/apluscourses/utils/DateDifferenceFormatter.java
package fi.aalto.cs.apluscourses.utils;

import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;

public class DateDifferenceFormatter {
  // the order of time units to check, ordered from most to least precise
  private static final List<ChronoUnit> TimeUnits = Arrays.asList(
          ChronoUnit.SECONDS,
          ChronoUnit.MINUTES,
          ChronoUnit.HOURS,
          ChronoUnit.DAYS,
          ChronoUnit.MONTHS,
          ChronoUnit.YEARS
  );

  private static String formatWithTimeUnit(ZonedDateTime earlierTimePoint,
                                           ZonedDateTime laterTimePoint,
                                           ChronoUnit timeUnit) {
    long timeUnitCount = timeUnit.between(earlierTimePoint, laterTimePoint);
    String timeUnitText = timeUnit.toString().toLowerCase(Locale.ROOT);
    if (timeUnitCount == 1) {
      // remove the final pluralizing "s" from the time unit name
      timeUnitText = timeUnitText.substring(0, timeUnitText.length() - 1);
    }

    return timeUnitCount + " " + timeUnitText + " ago";
  }

  /**
   * Returns a string containing a formatted time difference between two time points. The formatted
   * time is formatted in the ways of "3 seconds ago", "7 months ago" etc. Only the largest possible
   * time unit is used.
   *
   * @param earlierTimePoint The earlier time point.
   * @param laterTimePoint The later time point.
   */
  public static String formatWithLargestTimeUnit(ZonedDateTime earlierTimePoint,
                                                  ZonedDateTime laterTimePoint) {
    if (ChronoUnit.SECONDS.between(earlierTimePoint, laterTimePoint) <= 0) {
      return "just now";
    }

    for (int i = 1; i < TimeUnits.size(); i++) {
      if (TimeUnits.get(i).between(earlierTimePoint, laterTimePoint) == 0) {
        return formatWithTimeUnit(earlierTimePoint, laterTimePoint, TimeUnits.get(i - 1));
      }
    }

    return formatWithTimeUnit(earlierTimePoint, laterTimePoint,
            TimeUnits.get(TimeUnits.size() - 1));
  }

  private DateDifferenceFormatter() {

  }
}

File: src/main/kotlin/fi/aalto/cs/apluscourses/utils/temp/DateDifferenceFormatter.kt
package fi.aalto.cs.apluscourses.utils.temp

import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import kotlinx.datetime.TimeZone
import kotlinx.datetime.periodUntil
import java.time.temporal.ChronoUnit

object DateDifferenceFormatter {
    fun formatTimeUntilNow(time: Instant): String {
        val currentTime = Clock.System.now()
        return formatWithLargestTimeUnit(time, currentTime, true)
    }

    fun formatTimeSinceNow(time: Instant): String {
        val currentTime = Clock.System.now()
        return formatWithLargestTimeUnit(currentTime, time, false)
    }

    /**
     * Returns a string containing a formatted time difference between two time points. The formatted
     * time is formatted in the ways of "3 seconds ago", "7 months ago" etc. Only the largest possible
     * time unit is used.
     *
     * @param earlierTimePoint The earlier time point.
     * @param laterTimePoint   The later time point.
     */
    fun formatWithLargestTimeUnit(
        earlierTimePoint: Instant,
        laterTimePoint: Instant,
        before: Boolean
    ): String {
        val period = earlierTimePoint.periodUntil(laterTimePoint, TimeZone.UTC)

        if (period.seconds <= 0) {
            return "just now"
        }

        val (unit, time) = when {
            period.years > 0 -> ChronoUnit.YEARS to period.years
            period.months > 0 -> ChronoUnit.MONTHS to period.months
            period.days > 0 -> ChronoUnit.DAYS to period.days
            period.hours > 0 -> ChronoUnit.HOURS to period.hours
            period.minutes > 0 -> ChronoUnit.MINUTES to period.minutes
            else -> ChronoUnit.SECONDS to period.seconds
        }

        var timeUnitText = unit.toString().lowercase()
        if (time == 1) {
            // remove the final pluralizing "s" from the time unit name
            timeUnitText = timeUnitText.substringBeforeLast("s")
        }

        return if (before) {
            "$time $timeUnitText ago"
        } else {
            "in $time $timeUnitText"
        }
    }
}
